core/common/src/main/java/alluxio/AbstractClient.java

```diff
@@ -306,14 +311,45 @@ public synchronized void close() {
     mClosed = true;
   }
 
+  /**
+   *  {@link AbstractClient} works with Grpc Servers.
+   *  Child classes should only override this method to query the address
+   *  of the grpc server they talk to. The conversion from {@link GrpcServerAddress}
+   *  to more generic {@link SocketAddress} required by {@link Client#getRemoteSockAddress()}
+   *  is handled by this class.
+   *
+   * @return the {@link GrpcServerAddress} of the remote server
+   * @throws UnavailableException if address can't be determined
+   */
+  protected abstract GrpcServerAddress queryGrpcServerAddress() throws UnavailableException;
+
+  @Override
+  public SocketAddress getRemoteSockAddress() throws UnavailableException {
+    if (mServerAddress == null) {
+      mServerAddress = queryGrpcServerAddress();
+    }
+    return mServerAddress.getSocketAddress();
+  }
+
   @Override
-  public synchronized InetSocketAddress getAddress() throws UnavailableException {
-    return mAddress;
+  public String getRemoteHostName() throws UnavailableException {
+    if (mServerAddress == null) {
+      mServerAddress = queryGrpcServerAddress();
+    }
+    return mServerAddress.getHostName();
   }
 
   @Override
   public synchronized InetSocketAddress getConfAddress() throws UnavailableException {
-    return mAddress;
+    SocketAddress sockAddress = mServerAddress.getSocketAddress();
+    if (sockAddress instanceof InetSocketAddress) {
+      return (InetSocketAddress) sockAddress;
+    }
+
+    // We would reach here when the client talks to an Alluxio Worker through
+    // domain socket. Such client doesn't load configuration anyway. So this line
+    // currently should not be reached.
+    throw new UnavailableException("Remote is not an InetSockAddress");
   }
```

Comment:
This pr's scope is limited to refactoring the primary remote address APIs. So this is a work-around to retain the `getConfAddress` API that preserves the same behavior as before for all the master clients. For `BlockWorkClient` that uses a domain socket, this method will throw exception, but worker clients don't load configurations anyway.